state/protocol/protocol_state.go

```diff
@@ -0,0 +1,106 @@
+	// Entry Returns low-level protocol state entry that was used to initialize this object.
+	// It shouldn't be used by high-level logic, it is useful for some cases such as bootstrapping.
+	// Prefer using other methods to access protocol state.
+	Entry() *flow.RichProtocolStateEntry
+}
+
+// DynamicProtocolState extends the InitialProtocolState with data that can change from block to block.
+// It can be used to access the identity table at given block.
+type DynamicProtocolState interface {
+	InitialProtocolState
+	// EpochStatus returns the status of current epoch at given block based on the internal state of protocol.
+	EpochStatus() *flow.EpochStatus
+	// Identities returns identities that can participate in current and next epochs.
+	// Set of Authorized identities are different depending on epoch state:
+	// staking phase - identities for current epoch + identities from previous epoch (with 0 weight)
+	// setup & commit phase - identities for current epoch + identities from next epoch (with 0 weight)
+	Identities() flow.IdentityList
+	// GlobalParams returns params that are same for all nodes in the network.
+	GlobalParams() GlobalParams
+}
+
+// ProtocolState is the read-only interface for protocol state, it allows to query information
+// on a per-block and per-epoch basis.
+type ProtocolState interface {
+
+	// ByEpoch returns an object with static protocol state information by epoch number.
+	// To be able to use this interface we need to observe both epoch setup and commit events.
+	// Not available for next epoch unless we have observed an EpochCommit event.
+	// No errors are expected during normal operations.
+	// TODO(yuraolex): check return types
+	// TODO(yuraolex): decide if we really need this approach. It's unclear if it's useful to query
+	//  by epoch counter. To implement it we need an additional index by epoch counter. Alternatively we need a way to map
+	//  epoch counter -> block ID. It gets worse if we consider that we need a way to get the epoch counter itself at caller side.
+	//ByEpoch(epoch uint64) (InitialProtocolState, error)
+
+	// AtBlockID returns protocol state at block ID.
+	// The resulting protocol state is returned AFTER applying updates that are contained in block.
+	// Can be queried for any block that has been added to the block tree.
+	// Returns:
+	// - (DynamicProtocolState, nil) - if there is a protocol state associated with given block ID.
+	// - (nil, storage.ErrNotFound) - if there is no protocol state associated with given block ID.
+	// - (nil, exception) - any other error should be treated as exception.
+	AtBlockID(blockID flow.Identifier) (DynamicProtocolState, error)
+	// GlobalParams returns params that are the same for all nodes in the network.
+	GlobalParams() GlobalParams
+}
+
+// StateUpdater is a dedicated interface for updating protocol state.
+// It is used by the compliance layer to update protocol state when certain events that are stored in blocks are observed.
```

Comment:
Based on reading the current implementation, my interpretation is that the `ParticipantState` is responsible for validating input events, and only passing valid events to `StateUpdater`, then `StateUpdater` is responsible for computing the protocol state change associated with the event. 

It would be great to:
- Document this separation of responsibilities in the interface
- Since the `StateUpdater` methods only return irrecoverable errors, document that inputs must have already passed validation (eg. `isValidExtendingEpochSetup`)

Reply:
Added this to doc.